File: Infrastructure/Networking/vpc_manager.py
"""
VPC management and configuration
"""
import boto3
from typing import List, Dict


class VPCManager:
    """Manage AWS VPC resources"""
    
    def __init__(self, region: str = 'us-east-1'):
        self.ec2_client = boto3.client('ec2', region_name=region)
        self.region = region
    
    def create_vpc(self, cidr_block: str, name: str, enable_dns: bool = True):
        """Create VPC"""
        try:
            response = self.ec2_client.create_vpc(
                CidrBlock=cidr_block,
                AmazonProvidedIpv6CidrBlock=False
            )
            vpc_id = response['Vpc']['VpcId']
            
            # Enable DNS
            if enable_dns:
                self.ec2_client.modify_vpc_attribute(
                    VpcId=vpc_id,
                    EnableDnsHostnames={'Value': True}
                )
                self.ec2_client.modify_vpc_attribute(
                    VpcId=vpc_id,
                    EnableDnsSupport={'Value': True}
                )
            
            # Tag VPC
            self.ec2_client.create_tags(
                Resources=[vpc_id],
                Tags=[{'Key': 'Name', 'Value': name}]
            )
            
            return vpc_id
        except Exception as e:
            raise Exception(f"Failed to create VPC: {str(e)}")
    
    def create_subnet(self, vpc_id: str, cidr_block: str, az: str, 
                     name: str, public: bool = False):
        """Create subnet"""
        try:
            response = self.ec2_client.create_subnet(
                VpcId=vpc_id,
                CidrBlock=cidr_block,
                AvailabilityZone=az
            )
            subnet_id = response['Subnet']['SubnetId']
            
            # Enable auto-assign public IP for public subnets
            if public:
                self.ec2_client.modify_subnet_attribute(
                    SubnetId=subnet_id,
                    MapPublicIpOnLaunch={'Value': True}
                )
            
            # Tag subnet
            self.ec2_client.create_tags(
                Resources=[subnet_id],
                Tags=[
                    {'Key': 'Name', 'Value': name},
                    {'Key': 'Type', 'Value': 'Public' if public else 'Private'}
                ]
            )
            
            return subnet_id
        except Exception as e:
            raise Exception(f"Failed to create subnet: {str(e)}")
    
    def create_internet_gateway(self, vpc_id: str, name: str):
        """Create and attach internet gateway"""
        try:
            response = self.ec2_client.create_internet_gateway()
            igw_id = response['InternetGateway']['InternetGatewayId']
            
            # Attach to VPC
            self.ec2_client.attach_internet_gateway(
                InternetGatewayId=igw_id,
                VpcId=vpc_id
            )
            
            # Tag IGW
            self.ec2_client.create_tags(
                Resources=[igw_id],
                Tags=[{'Key': 'Name', 'Value': name}]
            )
            
            return igw_id
        except Exception as e:
            raise Exception(f"Failed to create internet gateway: {str(e)}")
    
    def create_nat_gateway(self, subnet_id: str, name: str):
        """Create NAT gateway"""
        try:
            # Allocate Elastic IP
            eip_response = self.ec2_client.allocate_address(Domain='vpc')
            allocation_id = eip_response['AllocationId']
            
            # Create NAT Gateway
            response = self.ec2_client.create_nat_gateway(
                SubnetId=subnet_id,
                AllocationId=allocation_id
            )
            nat_gw_id = response['NatGateway']['NatGatewayId']
            
            # Tag NAT Gateway
            self.ec2_client.create_tags(
                Resources=[nat_gw_id],
                Tags=[{'Key': 'Name', 'Value': name}]
            )
            
            return nat_gw_id
        except Exception as e:
            raise Exception(f"Failed to create NAT gateway: {str(e)}")
    
    def create_route_table(self, vpc_id: str, name: str, routes: List[Dict] = None):
        """Create route table"""
        try:
            response = self.ec2_client.create_route_table(VpcId=vpc_id)
            route_table_id = response['RouteTable']['RouteTableId']
            
            # Add routes
            if routes:
                for route in routes:
                    self.ec2_client.create_route(
                        RouteTableId=route_table_id,
                        **route
                    )
            
            # Tag route table
            self.ec2_client.create_tags(
                Resources=[route_table_id],
                Tags=[{'Key': 'Name', 'Value': name}]
            )
            
            return route_table_id
        except Exception as e:
            raise Exception(f"Failed to create route table: {str(e)}")

File: implementation/infrastructure-layer/network/vpc-manager.go
package main

import (
	"context"
	"fmt"
	"log"

	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/ec2"
	"github.com/aws/aws-sdk-go-v2/service/ec2/types"
)

type VPCManager struct {
	client *ec2.Client
	ctx    context.Context
}

func NewVPCManager(region string) (*VPCManager, error) {
	cfg, err := config.LoadDefaultConfig(context.TODO(), config.WithRegion(region))
	if err != nil {
		return nil, err
	}

	return &VPCManager{
		client: ec2.NewFromConfig(cfg),
		ctx:    context.Background(),
	}, nil
}

func (v *VPCManager) CreateVPC(cidrBlock string, name string) (string, error) {
	input := &ec2.CreateVpcInput{
		CidrBlock: &cidrBlock,
		TagSpecifications: []types.TagSpecification{
			{
				ResourceType: types.ResourceTypeVpc,
				Tags: []types.Tag{
					{
						Key:   stringPtr("Name"),
						Value: &name,
					},
				},
			},
		},
	}

	result, err := v.client.CreateVpc(v.ctx, input)
	if err != nil {
		return "", err
	}

	vpcID := *result.Vpc.VpcId
	log.Printf("VPC created: %s", vpcID)
	return vpcID, nil
}

func (v *VPCManager) CreateSubnet(vpcID, cidrBlock, az, name string) (string, error) {
	input := &ec2.CreateSubnetInput{
		VpcId:            &vpcID,
		CidrBlock:        &cidrBlock,
		AvailabilityZone: &az,
		TagSpecifications: []types.TagSpecification{
			{
				ResourceType: types.ResourceTypeSubnet,
				Tags: []types.Tag{
					{
						Key:   stringPtr("Name"),
						Value: &name,
					},
				},
			},
		},
	}

	result, err := v.client.CreateSubnet(v.ctx, input)
	if err != nil {
		return "", err
	}

	subnetID := *result.Subnet.SubnetId
	log.Printf("Subnet created: %s", subnetID)
	return subnetID, nil
}

func (v *VPCManager) CreateInternetGateway(name string) (string, error) {
	input := &ec2.CreateInternetGatewayInput{
		TagSpecifications: []types.TagSpecification{
			{
				ResourceType: types.ResourceTypeInternetGateway,
				Tags: []types.Tag{
					{
						Key:   stringPtr("Name"),
						Value: &name,
					},
				},
			},
		},
	}

	result, err := v.client.CreateInternetGateway(v.ctx, input)
	if err != nil {
		return "", err
	}

	igwID := *result.InternetGateway.InternetGatewayId
	log.Printf("Internet Gateway created: %s", igwID)
	return igwID, nil
}

func (v *VPCManager) AttachInternetGateway(vpcID, igwID string) error {
	input := &ec2.AttachInternetGatewayInput{
		VpcId:             &vpcID,
		InternetGatewayId: &igwID,
	}

	_, err := v.client.AttachInternetGateway(v.ctx, input)
	if err != nil {
		return err
	}

	log.Printf("Internet Gateway %s attached to VPC %s", igwID, vpcID)
	return nil
}

func stringPtr(s string) *string {
	return &s
}

func main() {
	manager, err := NewVPCManager("us-east-1")
	if err != nil {
		log.Fatal(err)
	}

	vpcID, err := manager.CreateVPC("10.0.0.0/16", "AutoCloud-VPC")
	if err != nil {
		log.Fatal(err)
	}

	fmt.Printf("VPC Created: %s\n", vpcID)
}
